pkg/types/urls_test.go

```diff
@@ -0,0 +1,181 @@
+				mustNewURL(t, "http://127.0.0.2:4001"),
+			},
+			"http://127.0.0.1:4001,http://127.0.0.2:4001",
+		},
+		{
+			URLs{
+				mustNewURL(t, "http://127.0.0.2:4001"),
+				mustNewURL(t, "http://127.0.0.1:4001"),
+			},
+			"http://127.0.0.2:4001,http://127.0.0.1:4001",
+		},
+	}
+	for i, tt := range tests {
+		g := tt.us.String()
+		if g != tt.wstr {
+			t.Errorf("#%d: string = %s, want %s", i, g, tt.wstr)
+		}
+	}
+}
+
+func TestURLsSort(t *testing.T) {
+	g := URLs{
+		mustNewURL(t, "http://127.0.0.4:4001"),
+		mustNewURL(t, "http://127.0.0.2:4001"),
+		mustNewURL(t, "http://127.0.0.1:4001"),
+		mustNewURL(t, "http://127.0.0.3:4001"),
+	}
+	w := URLs{
+		mustNewURL(t, "http://127.0.0.1:4001"),
+		mustNewURL(t, "http://127.0.0.2:4001"),
+		mustNewURL(t, "http://127.0.0.3:4001"),
+		mustNewURL(t, "http://127.0.0.4:4001"),
+	}
+	g.Sort()
+	if !reflect.DeepEqual(g, w) {
+		t.Errorf("URLs after sort = %#v, want %#v", g, w)
+	}
+}
+
+func TestURLsStringSlice(t *testing.T) {
+	tests := []struct {
+		us   URLs
+		wstr []string
+	}{
+		{
+			URLs{},
+			[]string{},
+		},
+		{
+			URLs{
+				mustNewURL(t, "http://127.0.0.1:4001"),
+			},
+			[]string{"http://127.0.0.1:4001"},
+		},
+		{
+			URLs{
+				mustNewURL(t, "http://127.0.0.1:4001"),
+				mustNewURL(t, "http://127.0.0.2:4001"),
+			},
+			[]string{"http://127.0.0.1:4001", "http://127.0.0.2:4001"},
+		},
+		{
+			URLs{
+				mustNewURL(t, "http://127.0.0.2:4001"),
+				mustNewURL(t, "http://127.0.0.1:4001"),
+			},
+			[]string{"http://127.0.0.2:4001", "http://127.0.0.1:4001"},
+		},
+	}
+	for i, tt := range tests {
+		g := tt.us.StringSlice()
+		if !reflect.DeepEqual(g, tt.wstr) {
+			t.Errorf("#%d: string slice = %+v, want %+v", i, g, tt.wstr)
+		}
+	}
+}
+
+func TestNewURLsFail(t *testing.T) {
+	tests := [][]string{
+		// no urls given
+		{},
+		// missing protocol scheme
+		{"://127.0.0.1:4001"},
+		// unsupported scheme
+		{"mailto://127.0.0.1:4001"},
+		// not conform to host:port
+		{"http://127.0.0.1"},
+		// contain a path
+		{"http://127.0.0.1:4001/path"},
+	}
+	for i, tt := range tests {
+		_, err := NewURLs(tt)
+		if err == nil {
+			t.Errorf("#%d: want error, but nil", i)
+		}
+	}
+}
+
+func TestNewURLs(t *testing.T) {
+	tests := []struct {
+		strs  []string
+		wurls URLs
+	}{
+		{
+			[]string{"http://127.0.0.1:4001"},
+			URLs{mustNewURL(t, "http://127.0.0.1:4001")},
+		},
+		// it can trim space
+		{
+			[]string{"   http://127.0.0.1:4001    "},
+			URLs{mustNewURL(t, "http://127.0.0.1:4001")},
+		},
+		// it does sort
+		{
+			[]string{
+				"http://127.0.0.2:4001",
+				"http://127.0.0.1:4001",
+			},
+			URLs{
+				mustNewURL(t, "http://127.0.0.1:4001"),
+				mustNewURL(t, "http://127.0.0.2:4001"),
+			},
+		},
+	}
+	for i, tt := range tests {
+		urls, _ := NewURLs(tt.strs)
+		if !reflect.DeepEqual(urls, tt.wurls) {
+			t.Errorf("#%d: urls = %+v, want %+v", i, urls, tt.wurls)
+		}
+	}
+}
+
+func mustNewURL(t *testing.T, s string) url.URL {
```

Comment:
obviously, it introduces a cyclone import then. pkg/testutil uses pkg/types/URLs, while pkg/types uses pkg/testutil to newURL. I think pkg/types is the root of the types, so i will keep mustNewURL here.